Yt/moist roe

slifer50 opened the pull request:
# Description

An Moist Roe flux implementation

<!--- Please fill out the following section --->

I have added a Moist Roe flux, a moist rising bubble test and Riemann shock tube type test. Also I have added additional boundary conditions necessary for running these test.

- [ ] Written and run all necessary tests with CLIMA by including `tests/runtests.jl`
- [ ] Followed all necessary [style guidelines](https://clima.github.io/ClimateMachine.jl/latest/DevDocs/CodeStyle/) and run `julia .dev/climaformat.jl .`
- [ ] Updated the documentation to reflect changes from this PR.

<!--- Please leave the following section --->

# For review by CLIMA developers

- [ ] There are no open pull requests for this already
- [ ] CLIMA developers with relevant expertise have been assigned to review this submission
- [ ] The code conforms to the [style guidelines](https://clima.github.io/ClimateMachine.jl/latest/DevDocs/CodeStyle/) and has consistent naming conventions. `julia .dev/format.jl` has been run in a separate commit.
- [ ] This code does what it is technically intended to do (all numerics make sense physically and/or computationally)

charleskawczynski commented on Project.toml:
```diff
@@ -23,6 +23,7 @@ Formatting = "59287772-0a20-5a39-b81b-1366585eb4c0"
 ForwardDiff = "f6369f11-7733-5829-9624-2563aa707210"
 GaussQuadrature = "d54b0c1a-921d-58e0-8e36-89d8069c0969"
 JLD2 = "033835bb-8acc-5ee8-8aae-3f567f8a3819"
+JuliaFormatter = "98e50ef6-434e-11e9-1051-2b60c6c9e899"
```
I think this was accidentally added. Let's remove this.

charleskawczynski commented on experiments/AtmosLES/bomex_model.jl:
```diff
@@ -511,6 +511,7 @@ function bomex_model(
         problem = problem,
         turbulence = SmagorinskyLilly{FT}(C_smag),
         moisture = moisture,
+        ref_state = NoReferenceState(),
```
This will likely result in CI failing. Probably need to revert this. If you need `NoReferenceState` for a specific experiment, maybe we can pass it in as a kwarg and set the `AtmosModel`'s default (`HydrostaticState(DecayingTemperatureProfile{FT}(param_set),)`).

charleskawczynski commented on experiments/AtmosLES/riemann.jl:
```diff
@@ -0,0 +1,345 @@
+    if (x < 0)
+        state.ρ = FT(1)
+        state.ρu = SVector{3, FT}(0, 0, 0)
+        p = FT(100000)
+        T = air_temperature_from_ideal_gas_law(param, p, FT(1))
+        state.ρe = internal_energy(param, T)
+        #state.moisture.ρq_tot = FT(0)
```
We should use a thermo state constructor here. We can add a new one, using `ρ` and `p`, if that's what you need and if it makes sense. Same for below.

charleskawczynski commented on src/Atmos/Model/AtmosModel.jl:
```diff
@@ -898,6 +906,264 @@ function numerical_flux_first_order!(
+    h̃ = RoeAverage(ρ⁻, ρ⁺, h⁻, h⁺)
+    qt = RoeAverage(ρ⁻, ρ⁺, qt⁻, qt⁺)
+    ρ = sqrt(ρ⁻ * ρ⁺)
+    e_int⁻ = internal_energy(
+        ρ⁻,
+        ρe⁻,
+        ρu⁻,
+        gravitational_potential(balance_law, state_auxiliary⁻),
+    )
```
```suggestion
    e_int⁻ = internal_energy(ts⁻)
```

charleskawczynski commented on src/Atmos/Model/AtmosModel.jl:
```diff
@@ -898,6 +906,264 @@ function numerical_flux_first_order!(
+        ρu⁻,
+        gravitational_potential(balance_law, state_auxiliary⁻),
+    )
+    e_int⁺ = internal_energy(
+        ρ⁺,
+        ρe⁺,
+        ρu⁺,
+        gravitational_potential(balance_law, state_auxiliary⁺),
+    )
```
```suggestion
    e_int⁺ = internal_energy(ts⁺)
```

charleskawczynski commented on src/Atmos/Model/AtmosModel.jl:
```diff
@@ -898,6 +906,264 @@ function numerical_flux_first_order!(
+    t,
+    direction,
+) where {S, A}
+
```
```suggestion
    balance_law.moisture isa EquilMoist || error("Must use a EquilMoist model for RoeNumericalFluxMoist")
```

charleskawczynski commented on src/Atmos/Model/AtmosModel.jl:
```diff
@@ -898,6 +906,264 @@ function numerical_flux_first_order!(
+        abs(ũᵀn),
+    )
+    if (numerical_flux.HH == true)
+        Λ = SDiagonal(
+            max(
+                abs(ũᵀn - c̃_LM),
+                max(
+                    0,
+                    ũᵀn - c̃_LM - (u⁻' * normal_vector - c⁻),
+                    u⁺' * normal_vector - c⁺ - (ũᵀn - c̃_LM),
+                ),
+            ),
```
Can we precompute these components to battle the formatter?

slifer50 replied:
This is written this way to be consistent with the formulation but yes, we could precompute this into separate quantities.

slifer50 commented:
@Zhengyu-Huang Can you have a look at the PR and let me know if there any additional changes necessary, I'm mainly trying to fix a documentation issue, would be good to know if there are any other things I should address.

charleskawczynski commented on src/Atmos/Model/AtmosModel.jl:
```diff
@@ -1142,4 +1150,251 @@ function numerical_flux_first_order!(
+    Δstate = SVector(Δρ, Δρu[1], Δρu[2], Δρu[3], Δρe, Δρq_tot)
+
+    parent(fluxᵀn) .-= M * Λ * (M \ Δstate) / 2
+    #@info parent(fluxᵀn)
```
```suggestion
```

charleskawczynski commented on src/Atmos/Model/linear.jl:
```diff
@@ -388,3 +388,134 @@ function numerical_flux_first_order!(
+    ΔpL = pL⁺ - pL⁻
+    Δρuᵀn = (ρu⁺ - ρu⁻)' * normal_vector
+
+    #fluxᵀn.ρ -= ΔpL / 2c̃
+    #fluxᵀn.ρu -= c̃ * Δρuᵀn * normal_vector / 2
+    #fluxᵀn.ρe -= h̃ * ΔpL / 2c̃
+
+    #temp1 = ΔpL / 2c̃
+    #temp2 = c̃ * Δρuᵀn * normal_vector / 2
+    #temp3 = h̃ * ΔpL / 2c̃
```
Should we delete this?

slifer50 replied:
Yes I'll get rid of this.

charleskawczynski commented on src/Atmos/Model/linear.jl:
```diff
@@ -388,3 +388,134 @@ function numerical_flux_first_order!(
+    _cv_d = cv_d(param_set)
+    Φ = gravitational_potential(balance_law.atmos, state_auxiliary⁻)
+    # guaranteed to be random
+    ω = Float64(π) / 3
+    δ = Float64(π) / 5
```
Is there a reason why we're hard-coding `Float64` here?

slifer50 replied:
I'll change this to an FT call.

charleskawczynski commented on src/Atmos/Model/linear.jl:
```diff
@@ -388,3 +388,134 @@ function numerical_flux_first_order!(
+    Δstate = SVector(Δρ, Δρu[1], Δρu[2], Δρu[3], Δρe, Δρq_tot)
+
+    parent(fluxᵀn) .-= M * Λ * (M \ Δstate) / 2
+    #temp = M * Λ * (M \ Δstate) / 2
+    #temp1 = ΔpL / 2c̃
+    #temp2 = c̃ * Δρuᵀn * normal_vector / 2
+    #temp3 = h̃ * ΔpL / 2c̃
+    #@info temp, temp1, temp2, temp3, state_prognostic⁺.moisture.ρq_tot, state_prognostic⁻.moisture.ρq_tot
```
Can we delete this if it's not being used?

slifer50 replied:
Yes, I'll get rid of this, it was just for testing.

slifer50 commented:
@charleskawczynski any idea what might be breaking documentation here? When I run make.jl on while connected to central I don't get any errors.

charleskawczynski commented on src/Atmos/Model/AtmosModel.jl:
```diff
@@ -1142,4 +1150,250 @@ function numerical_flux_first_order!(
+        balance_law.moisture.tolerance,
+    )
+    c̃ = sqrt(roe_average(ρ⁻, ρ⁺, c⁻^2, c⁺^2))
+    _cv_m = cv_m(ts)
+    R_m = gas_constant_air(ts)
+    _cp_m = cp_m(ts)
+    gamma = _cp_m / _cv_m
```
```suggestion
    (R_m, _cp_m, _cv_m, gamma) = gas_constants(ts)
```

Zhengyu-Huang commented on src/Atmos/Model/AtmosModel.jl:
```diff
@@ -1142,4 +1150,250 @@ function numerical_flux_first_order!(
+        SVector(1, ũ[1], ũ[2], ũ[3], ũ' * ũ / 2 + Φ - _T_0 * _cv_m, 0),
+        SVector(1, ũc̃⁻[1], ũc̃⁻[2], ũc̃⁻[3], h̃ + c̃ * ũᵀn, qt),
+        SVector(0, 0, 0, 0, _e_int_v0, 1),
+    )
```
Have you checked M * Λ * M^{-1} = d(F * n)/dU ? i.e., a unit test.  
This is not efficient (compared with the Dry Roe, OK to live with this now)

slifer50 replied:
Yes the Schur complement has to be calculated for us to make this as efficient as the DryRoe. We can do this at a later time. Would including the flux in the isentropic vortex test work for this purpose?

Zhengyu-Huang commented on src/Atmos/Model/AtmosModel.jl:
```diff
@@ -1142,4 +1150,250 @@ function numerical_flux_first_order!(
+    ω = FT(π) / 3
+    δ = FT(π) / 5
+    random_unit_vector = SVector(sin(ω) * cos(δ), cos(ω) * cos(δ), sin(δ))
+    # tangent space basis
```
Do any two tangential vectors τ1 τ2 work ? 
Do you need to add assert(  norm(τ1) > 0 ) ? I guess the random_unit_vector is to avoid random_unit_vector // normal_vector ?

slifer50 replied:
Yes pretty much, I don't think norm(τ1) = 0 is a possibility.

Zhengyu-Huang commented on src/Atmos/Model/linear.jl:
```diff
@@ -388,3 +388,119 @@ function numerical_flux_first_order!(
+    Φ = gravitational_potential(balance_law.atmos, state_auxiliary⁻)
+    # guaranteed to be random
+    ω = FT(π) / 3
+    δ = FT64(π) / 5
```
FT instead of FT64

Zhengyu-Huang commented on src/Atmos/Model/linear.jl:
```diff
@@ -388,3 +388,119 @@ function numerical_flux_first_order!(
+        state_auxiliary⁺,
+        t,
+        direction,
+    )
```
Is this a linearized flux ? why is this flux component nonlinear ? @mwarusz  This is also in the Dry roeflux

slifer50 replied:
I believe this uses the linearized pressure, can you point out which component is nonlinear?

Zhengyu-Huang commented on src/Atmos/Model/linear.jl:
```diff
@@ -388,3 +388,119 @@ function numerical_flux_first_order!(
+    ref_qliq⁻ = state_auxiliary⁻.ref_state.ρq_liq / ref_ρ⁻
+    ref_qice⁻ = state_auxiliary⁻.ref_state.ρq_ice / ref_ρ⁻
+    ref_p⁻ = state_auxiliary⁻.ref_state.p
+    #ref_h⁻ = (ref_ρe⁻ + ref_p⁻) / ref_ρ⁻
```
delete this line #.....

Zhengyu-Huang commented on src/Atmos/Model/linear.jl:
```diff
@@ -388,3 +388,119 @@ function numerical_flux_first_order!(
+    ref_qice⁻ = state_auxiliary⁻.ref_state.ρq_ice / ref_ρ⁻
+    ref_p⁻ = state_auxiliary⁻.ref_state.p
+    #ref_h⁻ = (ref_ρe⁻ + ref_p⁻) / ref_ρ⁻
+    q_pt⁻ = PhasePartition(ref_q⁻, ref_qliq⁻, ref_qice⁻)
```
is this  q_pt-     ref_q_pt-

Zhengyu-Huang commented on src/Atmos/Model/linear.jl:
```diff
@@ -388,3 +388,119 @@ function numerical_flux_first_order!(
+    ref_qliq⁺ = state_auxiliary⁺.ref_state.ρq_liq / ref_ρ⁺
+    ref_qice⁺ = state_auxiliary⁺.ref_state.ρq_ice / ref_ρ⁺
+    ref_p⁺ = state_auxiliary⁺.ref_state.p
+    #ref_h⁺ = (ref_ρe⁺ + ref_p⁺) / ref_ρ⁺
```
delete this line # .....

Zhengyu-Huang commented on src/Atmos/Model/linear.jl:
```diff
@@ -388,3 +388,119 @@ function numerical_flux_first_order!(
+    ref_qice⁺ = state_auxiliary⁺.ref_state.ρq_ice / ref_ρ⁺
+    ref_p⁺ = state_auxiliary⁺.ref_state.p
+    #ref_h⁺ = (ref_ρe⁺ + ref_p⁺) / ref_ρ⁺
+    q_pt⁺ = PhasePartition(ref_q⁺, ref_qliq⁺, ref_qice⁺)
```
q_pt+   ref_q_pt+

Zhengyu-Huang commented on src/Atmos/Model/linear.jl:
```diff
@@ -388,3 +388,119 @@ function numerical_flux_first_order!(
+    h̃ = (ref_h⁻ + ref_h⁺) / 2
+    c̃ = (ref_c⁻ + ref_c⁺) / 2
+    qt = (ref_q⁺ + ref_q⁻) / 2
+
```
add ref

Zhengyu-Huang reviewed (commented):
Great job! Thank you! But I have some (theoretical) questions.

slifer50 commented on src/Atmos/Model/AtmosModel.jl:
```diff
@@ -1142,4 +1150,250 @@ function numerical_flux_first_order!(
+        balance_law.moisture.maxiter,
+        balance_law.moisture.tolerance,
+    )
+    c̃ = sqrt(roe_average(ρ⁻, ρ⁺, c⁻^2, c⁺^2))
```
I'll adjust this to follow your suggestion

slifer50 commented on src/Numerics/DGMethods/NumericalFluxes.jl:
```diff
@@ -341,6 +342,32 @@ Requires a custom implementation for the balance law.
 """
 struct HLLCNumericalFlux <: NumericalFluxFirstOrder end
 
+"""
```
as far as I know the only reference we have for the moist Roe flux is our own documentation thanks to the work by @bischtob and @mwarusz, should we try to list this?

slifer50 commented:
@Zhengyu-Huang I think I've addressed your comments, can you have a look and let me know if there's anything I missed. I've added the flux to the isentropicvortex test. Do you have any other test suggestions?

kpamnany commented on src/Numerics/DGMethods/NumericalFluxes.jl:
```diff
@@ -363,6 +390,7 @@ An optional method can also be defined for
                                        Qaux⁻, Q⁺, Qstate_gradient_flux⁺, Qaux⁺, bctype, t)
 
 """
+
```
```suggestion
```

kpamnany commented on src/Atmos/Model/AtmosModel.jl:
```diff
@@ -1141,4 +1149,246 @@ function numerical_flux_first_order!(
+        Λ = SDiagonal(qa2, qa1, qa1, qa1, qa3, qa1)
+    end
+
+
+
+
```
```suggestion
```

kpamnany commented on src/Numerics/DGMethods/NumericalFluxes.jl:
```diff
@@ -341,6 +342,32 @@ Requires a custom implementation for the balance law.
+
+Requires a custom implementation for the balance law.
+"""
+
```
```suggestion
```

slifer50 commented:
@Zhengyu-Huang @charleskawczynski @mwarusz I think this is about ready to be merged, are there any concerns that still need to be addressed before we can r+ this?

Zhengyu-Huang reviewed (approved):
That looks great! Thank you! Like the new test!

slifer50 commented:
Yes this should apply the central flux the linearized balance law

slifer50 commented:
I've just rebased this, @charleskawczynski @Zhengyu-Huang let's r+ if there are no objections

charleskawczynski commented on test/Numerics/DGMethods/Euler/isentropicvortex.jl:
```diff
@@ -222,15 +243,19 @@ function test_run(
         ref_state = NoReferenceState(),
         turbulence = ConstantDynamicViscosity(FT(0)),
-        moisture = DryModel(),
+        moisture = moisture,#DryModel(),
```
```suggestion
        moisture = moisture,
```

charleskawczynski commented on src/Atmos/Model/linear.jl:
```diff
@@ -388,3 +388,124 @@ function numerical_flux_first_order!(
+    Δρuᵀn = (ρu⁺ - ρu⁻)' * normal_vector
+
+    _T_0::FT = T_0(param_set)
+    _e_int_v0 = e_int_v0(param_set)
+    _cv_d = cv_d(param_set)
```
```suggestion
    _e_int_v0::FT = e_int_v0(param_set)
    _cv_d::FT = cv_d(param_set)
```

charleskawczynski commented on src/Atmos/Model/AtmosModel.jl:
```diff
@@ -1141,4 +1149,243 @@ function numerical_flux_first_order!(
+    _cv_d::FT = cv_d(param_set)
+    _T_0::FT = T_0(param_set)
+    γ::FT = cp_d(param_set) / cv_d(param_set)
+    _e_int_v0 = e_int_v0(param_set)
```
```suggestion
    _e_int_v0::FT = e_int_v0(param_set)
```

charleskawczynski commented on src/Atmos/Model/AtmosModel.jl:
```diff
@@ -1141,4 +1149,243 @@ function numerical_flux_first_order!(
+        balance_law.moisture.maxiter,
+        balance_law.moisture.tolerance,
+    )
+    c̃ = sqrt((γ - 1) * (h̃ - (ũ[1]^2 + ũ[2]^2 + ũ[3]^2) / 2))#sqrt(roe_average(ρ⁻, ρ⁺, c⁻^2, c⁺^2))
```
```suggestion
    c̃ = sqrt((γ - 1) * (h̃ - (ũ[1]^2 + ũ[2]^2 + ũ[3]^2) / 2))
```

charleskawczynski commented on src/Atmos/Model/AtmosModel.jl:
```diff
@@ -1141,4 +1149,243 @@ function numerical_flux_first_order!(
+    if (numerical_flux.LM == true)
+        Mach⁺ = sqrt(u⁺' * u⁺) / c⁺
+        Mach⁻ = sqrt(u⁻' * u⁻) / c⁻
+        Mach = (Mach⁺ + Mach⁻) / 2 #roe_average(ρ⁻, ρ⁺, Mach⁻, Mach⁺)
```
```suggestion
        Mach = (Mach⁺ + Mach⁻) / 2
```

slifer50 commented:
I've rebased this again.

charleskawczynski reviewed (commented):
Can we add tests for when `LM, HH, LV, LVPP` are true (I'm assuming one at a time?)

slifer50 commented:
> Can we add tests for when `LM, HH, LV, LVPP` are true (I'm assuming one at a time?)

I can try to include this in the isentropic vortex test, I'll get to work on it.

slifer50 commented:
@charleskawczynski I've added in the LM, etc tests to the isentropic vortex case.

slifer50 commented:
Is this in a mergeable state ? @charleskawczynski

charleskawczynski commented on test/Numerics/DGMethods/Euler/isentropicvortex.jl:
```diff
@@ -379,6 +498,9 @@ function isentropicvortex_initialcondition!(
     state.ρu = ρ * u
     e_kin = u' * u / 2
     state.ρe = ρ * total_energy(e_kin, e_pot, ts)
+    if bl.moisture isa EquilMoist
```
```suggestion
    if !(bl.moisture isa DryModel)
```

charleskawczynski commented on test/Numerics/DGMethods/Euler/isentropicvortex.jl:
```diff
@@ -222,22 +337,26 @@ function test_run(
         model,
         grid,
-        NumericalFlux(),
+        NumericalFlux,#(),
```
```suggestion
        NumericalFlux,
```

slifer50 commented:
rebase?

charleskawczynski reviewed (approved):
LGTM. Let's squash and merge?

slifer50 commented:
Squashed and good to go

charleskawczynski commented:
bors r+

charleskawczynski commented on src/Atmos/Model/AtmosModel.jl:
```diff
@@ -1141,4 +1149,243 @@ function numerical_flux_first_order!(
+        abs(ũᵀn),
+    )
+    #Harten Hyman
+    if (numerical_flux.HH == true)
```
@slifer50 just to note, it might be an improvement to use dispatch for these, rather than `Bool`s. But the method interface might not be the same and some parts are needed for multiple numerical fluxes, so I couldn't see a clear win at first glance.

slifer50 replied:
Yeah, I'm not a big fan of the current way this is being handled, but I don't really know a simple way to do this without building these as entirely separate fluxes.

charleskawczynski replied:
We can try to do that, if it's worthwhile, in a follow-up PR

slifer50 replied:
I agree with this.
